state/src/replicationv2/mod.rs

```diff
@@ -273,51 +311,265 @@ mod test {
 
         // Check a random node's DB to ensure the wallet exists
         let nid = (0..N).choose(&mut rng).unwrap();
-        let tx = nodes[nid].get_db().new_read_tx().unwrap();
+        let db = nodes.get_db(nid as NodeId).await;
+        let tx = db.new_read_tx().unwrap();
         let wallet = tx.get_wallet(&wallet_id).unwrap();
         assert!(wallet.is_some());
     }
 
-    #[tokio::test(flavor = "multi_thread", worker_threads = 5)]
+    /// Tests proposing a state transition to a random node, assuming it will be
+    /// forwarded to the leader
+    #[tokio::test]
     #[allow(non_snake_case)]
     async fn test_simple_raft__propose_random() {
         const N: usize = 5;
         let mut rng = thread_rng();
 
         // Setup a raft
         let nodes = MockRaft::create_raft(N).await;
-        let leader = nodes[0].get_client().leader().await;
+        let leader = nodes.get_client(0).await.leader().await;
         if leader.is_none() {
             panic!("no leader in raft");
         }
 
         // Propose a wallet update to a random node
         let nid = (0..N).choose(&mut rng).unwrap();
-        let target_raft = nodes[nid].get_client();
+        let target_raft = nodes.get_client(nid as NodeId).await;
         let wallet = mock_empty_wallet();
         let wallet_id = wallet.wallet_id;
         target_raft.propose_transition(StateTransition::AddWallet { wallet }).await.unwrap();
         tokio::time::sleep(Duration::from_millis(100)).await;
 
         // Check a random node's DB to ensure the wallet exists
         let nid = (0..N).choose(&mut rng).unwrap();
-        let tx = nodes[nid].get_db().new_read_tx().unwrap();
+        let db = nodes.get_db(nid as NodeId).await;
+        let tx = db.new_read_tx().unwrap();
+        let wallet = tx.get_wallet(&wallet_id).unwrap();
+        assert!(wallet.is_some());
+    }
+
+    /// Tests adding a new node to the raft
+    #[tokio::test]
+    #[allow(non_snake_case)]
+    async fn test_add_node__to_singleton() {
+        let mut rng = thread_rng();
+
+        // Begin a singleton cluster
+        let raft = MockRaft::create_singleton_raft().await;
+        let client = raft.get_client(0).await;
+        let wallet = mock_empty_wallet();
+        let wallet_id = wallet.wallet_id;
+        client.propose_transition(StateTransition::AddWallet { wallet }).await.unwrap();
+
+        // Add a new node
+        let new_nid = 2;
+        raft.add_node(new_nid).await;
+
+        // Add the node as a learner
+        client.add_learner(new_nid).await.unwrap();
+        tokio::time::sleep(Duration::from_millis(100)).await;
+
+        // Check the DB of the new node
+        let db = raft.get_db(new_nid).await;
+        let tx = db.new_read_tx().unwrap();
         let wallet = tx.get_wallet(&wallet_id).unwrap();
         assert!(wallet.is_some());
+
+        // Promote the learner, propose a new wallet, ensure consensus is reached
+        client.promote_learner(new_nid).await.unwrap();
```

Comment:
is it important for the new learner to be promoted for these tests? it should receive new proposals regardless, right? which leads me to the question: in these join/leave tests, would it be helpful to test that after promotion, the new node can successfully make new proposals of its own?

Reply:
To your first question: it does not matter whether the node is a learner or voter for this test

To your second: Technically only the leader can create proposals. We just do the extra step of having each client forward proposals to the leader. So it doesn't matter whether the node is a follower or learner here either.